xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,362 @@
+import io.grpc.netty.InternalProtocolNegotiator;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.ConnectionSourceType;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.util.AsciiString;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger log = Logger.getLogger(FilterChainMatchingHandler.class.getName());
+
+  private static final AsciiString SCHEME = AsciiString.of("http");
+
+  private final GrpcHttp2ConnectionHandler grpcHandler;
+  private final FilterChainSelector selector;
+  private final ProtocolNegotiator delegate;
+
+  FilterChainMatchingHandler(
+          GrpcHttp2ConnectionHandler grpcHandler, FilterChainSelector selector,
+          ProtocolNegotiator delegate) {
+    this.grpcHandler = checkNotNull(grpcHandler, "grpcHandler");
+    this.selector = checkNotNull(selector, "selector");
+    this.delegate = checkNotNull(delegate, "delegate");
+  }
+
+  @Override
+  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    if (!(evt instanceof ProtocolNegotiationEvent)) {
+      super.userEventTriggered(ctx, evt);
+      return;
+    }
+    SelectedConfig config = selector.select(
+            (InetSocketAddress) ctx.channel().localAddress(),
+            (InetSocketAddress) ctx.channel().remoteAddress());
+    if (config == null) {
+      log.log(Level.FINER, "Did not find exactly one filter chain.");
+      ctx.fireExceptionCaught(
+              new IllegalStateException("Did not find exactly one filter chain."));
+      return;
+    }
+    ProtocolNegotiationEvent pne = (ProtocolNegotiationEvent)evt;
+    Attributes attr = InternalProtocolNegotiationEvent.getAttributes(pne)
+            .toBuilder().set(ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER,
+                    config.sslContextProviderSupplier).build();
+    pne = InternalProtocolNegotiationEvent.withAttributes(pne, attr);
+    ctx.pipeline().replace(this, null, delegate.newHandler(grpcHandler));
```

Comment:
Something I noticed last time too but forgot to comment. I am a bit troubled by the generic use of a `delegate` and `delegate.newHandler(grpcHandler)` but then setting the security specific `ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER` based on the assumption that the delegate's handler is going to consume it so the delegate has to be `ServerSdsProtocolNegotiator`.

One way to address this is to replace `ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER` with say `ATTR_SERVER_SELECTED_CONFIG` and then set `config` as the value of that attribute. Then the delegate's handler can get the `config` and get any of the values in that config - currently only `sslContextProviderSupplier` but in future it could be other things from the selected config. Of course if we are only ever going to have security and routing config as selected config from the filter-chain then we don't need the generic logic.

Reply:
It is up to the delegate to decide how to consume the `config` we set here. It is currently only `ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER ` but by-design we would add routing config, maybe in one data structure. 

The point of the delegate is to saying loudly that the `xdsServerCredential`/`ServerSdsProtocolNegotiator` path is separate from the `filterChainMatchingProtocolNegotiators`, the former being the next step.

Reply:
Your comment doesn't answer my question so let me restate it a bit differently:

instead of doing `.set(ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER, config.sslContextProviderSupplier)` would it be better to do, say, `.set(ATTR_SERVER_SELECTED_CONFIG, config)` and then modify the other code currently using ATTR_SERVER_SSL_CONTEXT_PROVIDER_SUPPLIER to instead use ATTR_SERVER_SELECTED_CONFIG and then get the member values out of `config` (which is a `SelectedConfig`)?

Reply:
That is what we are going to do immediately, what about we do that in https://github.com/grpc/grpc-java/pull/8318? I know your suggestion probably sets the boundary between the two PRs more clearly, but since eventually it would be the same when in release it should be fine, and it makes it much easier for me.